### PR TITLE
Introduce the structure that will hold derivatives for the Newton method

### PR DESCRIPTION
**Pull request by bangerth**
I'm posting this patch for @MFraters out of whose repository I have this patch. I committed it in his name.

This introduces the additional data structures necessary to hold the derivatives that we will need for the Newton method on the Stokes system. It also has a class `NewtonHandler` similar to the `MeltHandler` class. Comments welcome.

### CODE REVIEW

File: include/aspect/newton.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+namespace aspect
+{
+  using namespace dealii;
+
+  namespace MaterialModel
+  {
+
+    template <int dim>
+    class MaterialModelDerivatives : public AdditionalMaterialOutputs<dim>
```
Comment (tjhei) [inline include/aspect/newton.h]: document the class please

Comment (tjhei) [inline include/aspect/newton.h]: Are these going to be all derivatives in the future because right now they are only viscosity derivatives for the Stokes block. If this stays like this, it should be reflected in the name.

File: include/aspect/newton.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    {
+      public:
+        /**
+         * Constructor. Initialize the various arrays of this structure with the
+         * given number of quadrature points and (finite element) components.
+         *
+         * @param n_points The number of quadrature points for which input
+         * quantities will be provided.
+         * @param n_comp The number of vector quantities (in the order in which
```
Comment (tjhei) [inline include/aspect/newton.h]: This is the number of compositions? Remove the documentation here because it is misleading.

File: include/aspect/newton.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+          dviscosities_dvelocity.resize(n_points, numbers::signaling_nan<double>());
+          dviscosities_dpressure.resize(n_points, numbers::signaling_nan<double>());
+          dviscosities_dtemperature.resize(n_points, numbers::signaling_nan<double>());
+          dviscosities_dstrain_rate.resize(n_points, numbers::signaling_nan<SymmetricTensor<2,dim> >());
+        };
+
+        /**
+         * The derivatives of the viscosities
+         */
+        std::vector<double> dviscosities_dvelocity;
```
Comment (tjhei) [inline include/aspect/newton.h]: I would prefer something like ``d_viscosities_d_velocity`` or ``viscosity_derivative_wrt_velocity`` or ``partial_velocity_viscosities``.

File: include/aspect/newton.h [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        std::vector<std::vector<double> > dviscosities_dcompositions;
+
+    };
+  }
+
+  /**
+   * Class that contains all runtime parameters and other helper functions
+   * related to Newton solver. A global instance can be retrieved with
+   * SimulatorAccess<dim>::get_newton_handler(), but keep in mind that it only
+   * exists if parameters.include_melt_transport is true. TODO: rewrite
```
Comment (tjhei) [inline include/aspect/newton.h]: ;-)

File: include/aspect/newton.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+   */
+  template <int dim>
+  class NewtonHandler: public SimulatorAccess<dim>
```
Comment (tjhei) [inline include/aspect/newton.h]: Is the plan to have a ``NewtonHandler`` like ``MeltHandler``? I think this won't be necessary.

File: include/aspect/newton.h (hunk 1 above)
Comment (MFraters) [inline include/aspect/newton.h]: Yes, at least, that was what my idea was for this class to also store derivatives related to the material model. For now, I think we could even get rid of dviscosities_dvelocity and dviscosities_dtemperature, because they are not used at this moment.

File: include/aspect/newton.h (hunk 5 above)
Comment (MFraters) [inline include/aspect/newton.h]: Back when I had to rewrite the assembly into the new style, I looked at the options and this seemed to be the best and cleanest way to put it in the new assembly framework. So where do you propose to put it?

File: include/aspect/newton.h [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+         * @param n_comp The number of vector quantities (in the order in which
+         * the Introspection class reports them) for which input will be
+         * provided.
+         */
+        MaterialModelDerivatives (const unsigned int n_points,
+                                  const unsigned int /*n_comp*/)
+        {
+          dviscosities_dvelocity.resize(n_points, numbers::signaling_nan<double>());
+          dviscosities_dpressure.resize(n_points, numbers::signaling_nan<double>());
+          dviscosities_dtemperature.resize(n_points, numbers::signaling_nan<double>());
```
Comment (tjhei) [inline include/aspect/newton.h]: Why do we have the derivative wrt T but not composition?

File: include/aspect/newton.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+          dviscosities_dstrain_rate.resize(n_points, numbers::signaling_nan<SymmetricTensor<2,dim> >());
```
Comment (tjhei) [inline include/aspect/newton.h]: you are not initializing ``dviscosities_dcompositions``.

**Comment by bangerth** [comment]
@MFraters -- We had a conversation here about how best to address review comments. I'm happy to make the changes that were requested, or I could leave it to you and then just extract the modified files and make them into another pull request. Either is fine by me.

One way or the other, it is going to cause you some pain if you want to keep your branch up to date. There is of course the option that you simply don't want to keep your branch up to date, and that we keep merging pieces. Or that I extract things from your branch, fix them up in commits of my own, and then all you have to do is cherry-pick my own commits from mainline to keep your branch (mostly) up to date. Or that you frequently merge mainline into your branch until, hopefully at one point, your branch will no longer contain any changes that have not already been merged.

I think my preference is that we proceed with taking small parts of your work at a time. The experience from the particle work by @gassmoeller and the melt work by @jdannberg and @tjhei taught me that the idea of reviewing multi-1000 lines of code is simply too painful for everyone. Let's rather do things one step at a time.

What do you think about these issues?

**Comment by MFraters** [comment]
@bangerth I think for the sake of fast and efficient progress I am fine with you (or Timo) making the changes which you request. But I would of course like to review chose changes before they are committed to the main repository, and I would also like to know the reason why you want to do things differently. If you rather have me making the changes (or a specific part of them) I am of course happy to do that.

Since you have way more experience in reviewing large chunks of code, I will leave that up to you. But the problem of cutting it up into small parts may be that you can't always see how the choices that you make in one part will affect the others, while doing it in one go, I could check if all the benchmarks still work. Having said that, I guess the review process of such a large chunk will be horrible.

**Comment by bangerth** [comment]
@MFraters -- the reason for the procedure is that I've learned that reviewing very large patches is just not practical. Both the patches mentioned above had been sitting on github for weeks or months before I could find the time to go through their entirety, simply because (i) I dreaded the enormity of the task, and (ii) I could simply not find enough consecutive time to make it happen. For both patches, I had to set aside an entire day or more to read through the whole thing, and then do it again after revisions. I believe @gassmoeller received more than 200 emails that day with comments from me, and it took him more than 2 weeks before all were addressed. I think the overall numbers may have been even larger for the melt functionality.

In other words, I've come to conclude that this approach is simply not efficient -- it leads to long delays because all of the work has to be done at once. I think a piecemeal approach may in fact be faster. It clearly has its drawbacks, but I would like to try it to see whether we can get more eyes on the problem if we split it up into smaller chunks.

**Comment by bangerth** [comment]
And "more eyes" -> "faster turnaround".

File: include/aspect/newton.h (hunk 5 above)
Comment (gassmoeller) [inline include/aspect/newton.h]: @tjhei was it not necessary to have a class that implements the `create_material_model_outputs` function? I remember something like this from the melt implementation.

File: include/aspect/newton.h [hunk 8]
```diff
@@ -0,0 +1,116 @@
+/*
+  Copyright (C) 2016 by the authors of the ASPECT code.
```
Comment (gassmoeller) [inline include/aspect/newton.h]: 2017

File: source/simulator/newton.cc [hunk 9]
```diff
@@ -0,0 +1,85 @@
+/*
+  Copyright (C) 2016 by the authors of the ASPECT code.
```
Comment (gassmoeller) [inline source/simulator/newton.cc]: 2017

**Comment by bangerth** [comment]
Superseded by #1499.
